test/FunctionalTests/Client/UnaryTests.cs

```diff
@@ -68,6 +68,35 @@ Task<HelloReply> UnaryThrowError(HelloRequest request, ServerCallContext context
         StringAssert.StartsWith("Failed to deserialize response message.", call.GetStatus().Detail);
     }
 
+    [Test]
+    public async Task WriteResponseHeadersAsync_HeadersSentEarly()
+    {
+        var tcs = new TaskCompletionSource<HelloReply>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        async Task<HelloReply> UnaryThrowError(HelloRequest request, ServerCallContext context)
+        {
+            await context.WriteResponseHeadersAsync(new Metadata
+            {
+                new Metadata.Entry("key", "value")
+            });
+
+            return await tcs.Task;
+        }
+
+        // Arrange
+        var method = Fixture.DynamicGrpc.AddUnaryMethod<HelloRequest, HelloReply>(UnaryThrowError);
+        var channel = CreateChannel();
+        var client = TestClientFactory.Create(channel, method);
+
+        // Act
+        var call = client.UnaryCall(new HelloRequest());
+
+        // Assert
+        var headers = await call.ResponseHeadersAsync.DefaultTimeout();
+        Assert.AreEqual("value", headers.GetValue("key"));
+        Assert.IsFalse(call.ResponseAsync.IsCompleted);
+    }
```

Comment:
super nit so you don't make the GC clean up the task:
```suggestion
        tcs.SetResult(new HelloReply() { Message = "" });
    }
```